geoh5py/ui_json/forms.py

```diff
@@ -0,0 +1,499 @@
+    :param tooltip: String rendered on hover over ui element.
+
+    :note: Standardized form members are accessible through public namespace
+        by way of the ValueAccess descriptor.
+    """
+
+    validations = {"required_form_members": ["label", "value"]}
+    identifier_members: list[str] = []
+
+    def __init__(
+        self,
+        name: str,
+        value: Any | None = None,
+        **kwargs,
+    ):
+        self.name: str = name
+        self._value: Parameter = self._set_value_parameter(value)
+        self._label = StringParameter("label")
+        self._enabled = BoolParameter("enabled", value=True)
+        self._optional = BoolParameter("optional", value=False)
+        self._group_optional = BoolParameter("group_optional", value=False)
+        self._main = BoolParameter("main", value=True)
+        self._group = StringParameter("group")
+        self._dependency = StringParameter("dependency")
+        self._dependency_type = StringParameter("dependency_type", value="enabled")
+        self._group_dependency = StringParameter("group_dependency")
+        self._group_dependency_type = StringParameter(
+            "group_dependency_type", value="enabled"
+        )
+        self._tooltip = StringParameter("tooltip")
+        self._allow_values_access()
+        self._extra_members: dict[str, Any] = {}
+        self._active_members: list[str] = []
+        if kwargs:
+            self.register(kwargs)
+
+    def form(self, use_camel: bool = False) -> dict[str, Any]:
+        """
+        Returns dictionary of active form members and their values.
+
+        :param use_camel: If True, keys are converted to camel case.
+        """
+        form = {}
+        for member in self.active:
+            if member in self._extra_members:
+                form[member] = self._extra_members[member]
+            else:
+                form[member] = getattr(self, member)
+
+        if use_camel:
+            form = MEMBER_KEYS.map(form, "camel")
+
+        return form
+
+    def register(self, members: dict[str, Any]):
+        """
+        Update form members with incoming values.
+
+        :param members: Dictionary of form members and associated data.
+        """
+
+        error_list = []
+        members = MEMBER_KEYS.map(members)
+        for member in list(members):
+            if member in self.valid_members:
+                try:
+                    setattr(self, member, members.pop(member))
+                except BaseValidationError as err:
+                    error_list.append(err)
+
+        if error_list:
+            if len(error_list) == 1:
+                raise error_list.pop()
+            raise AggregateValidationError(self.name, error_list)
+
+        self._extra_members.update(members)
+
+    def validate(self):
+        """Validates data against the pool of enforcers."""
+        enforcers = EnforcerPool.from_validations(self.name, self.validations)
+        enforcers.validate(self.form())

```

Comment:
Is this old? I don't see a validate method on the enforcer.

Reply:
Ok now I get that the method is named different on the Pool. Maybe we keep the name `enforce`? We could even just make it a `__call__(self)`

Reply:
renamed

Reply:
using the __call__ method here makes the code less readable.  It's more obvious what's going on if we just use `enforcers.enforce`.